src/corelibs/U2View/src/ov_msa/AlignSequencesToAlignment/AlignSequencesToAlignmentSupport.cpp

```diff
@@ -0,0 +1,126 @@
+    : GObjectViewWindowContext(parent, MsaEditorFactory::ID) {
+}
+
+void AlignSequencesToAlignmentSupport::initViewContext(GObjectView *view) {
+    auto msaEditor = qobject_cast<MSAEditor *>(view);
+    SAFE_POINT(msaEditor != nullptr, "View is not MSAEditor!", );
+    CHECK(msaEditor->getMaObject() != nullptr, );
+
+    // TODO: using a single hardcoded (MAFFT) today (this is the original UGENE behavior).
+    //  The final code should build action per each algorithm registered in the AlignmentAlgorithmsRegistry.
+    auto alignAction = new AlignSequencesToAlignmentAction(this, msaEditor, tr("Add sequences with MAFFT..."), 100);
+    alignAction->setIcon(QIcon(":/core/images/add_to_alignment.png"));    //TODO: add a dedicated icon for MAFFT.
+    alignAction->setObjectName("Align with MAFFT");
+    alignAction->setMenuTypes({MsaEditorMenuType::ALIGN_SEQUENCES_TO_ALIGNMENT});
+    alignAction->sl_updateState();
+
+    connect(alignAction, SIGNAL(triggered()), SLOT(sl_alignSequencesToAlignment()));
+    addViewAction(alignAction);
+}
+
+void AlignSequencesToAlignmentSupport::sl_alignSequencesToAlignment() {
+    auto action = qobject_cast<AlignSequencesToAlignmentAction *>(sender());
+    SAFE_POINT(action != nullptr, "Not a AlignSequencesToAlignmentAction!", );
+    auto msaEditor = action->getEditor();
+    auto msaObject = msaEditor->getMaObject();
+    CHECK(!msaObject->isStateLocked(), );
+
+    ProjectView *projectView = AppContext::getProjectView();
+    SAFE_POINT(projectView != nullptr, "Project view is null", );
+
+    const GObjectSelection *selection = projectView->getGObjectSelection();
+    SAFE_POINT(selection != nullptr, "GObjectSelection is null", );
+
+    QList<GObject *> objects = selection->getSelectedObjects();
+    bool selectFromProject = !objects.isEmpty();
+
+    for (const GObject *object : qAsConst(objects)) {
+        const GObjectType &objectType = object->getGObjectType();
+        if (object == msaObject || (objectType != GObjectTypes::MULTIPLE_SEQUENCE_ALIGNMENT && objectType != GObjectTypes::SEQUENCE)) {
+            selectFromProject = false;
+            break;
+        }
+    }
+
+    // TODO: ask user if to select from the project or from files.
+    if (selectFromProject) {
+        SequenceObjectsExtractor extractor;
+        extractor.setAlphabet(msaObject->getAlphabet());
+        extractor.extractSequencesFromObjects(objects);
+
+        if (!extractor.getSequenceRefs().isEmpty()) {
+            auto task = new AlignSequencesToAlignmentTask(msaObject, extractor);
+            TaskWatchdog::trackResourceExistence(msaObject, task, tr("A problem occurred during adding sequences. The multiple alignment is no more available."));
+            AppContext::getTaskScheduler()->registerTopLevelTask(task);
+        }
+    } else {
+        QString filter = DialogUtils::prepareDocumentsFileFilterByObjType(GObjectTypes::SEQUENCE, true);
```

Comment:
It's possible to choose alignments from the project, but here sequences only are available? I think it's better to unify it.

Reply:
Same here: out of scope of this PR: this is an old behavior copied into the new cpp file out of MSAEditor.cpp.